static/js/controllers/contacts-edit.js

```diff
@@ -192,142 +193,121 @@ angular.module('inboxControllers').controller('ContactsEditCtrl',
+            submitted.doc.type = $scope.enketoContact.type;
           }
 
-          return saveDoc(submitted, original, extras, repeated);
+          return prepareSubmittedDocsForSave(original, submitted);
+        })
+        .then(function(preparedDocs) {
+          return DB().bulkDocs(preparedDocs.preparedDocs)
+            .then(function(bulkDocResult) {
+              if (_.some(bulkDocResult, function(result) {
+                return !result.ok;
+              })) {
+                // REVIEWER: is this OK? What would you throw here. It doesn't
+                //           seem feasible to throw a generic error message and
+                //           then attach the result in a meaningful way.
+                throw bulkDocResult;
+              }
+
+              return {docId: preparedDocs.docId, bulkDocsResult: bulkDocResult};
+            });
         });
     }
 
-    function persist(doc) {
-      updateTitle(doc);
-
-      var put;
-      if(doc._id) {
-        put = DB().put(doc);
-      } else {
+    // Prepares document to be bulk-saved at a later time, and for it to be
+    // referenced by _id by other docs if required.
+    function prepare(doc) {
+      if(!doc._id) {
         doc.reported_date = Date.now();
-        put = DB().post(doc);
+        doc._id = uuidV1();
       }
-      return put
-        .then(function(response) {
-          return DB().get(response.id);
-        });
-    }
 
-    function saveRepeated(repeated) {
-      if (!(repeated && repeated.child_data)) {
-        return $q.resolve();
-      }
+      return doc;
+    }
 
-      var children = [];
-      return $q
-        .all(_.map(repeated.child_data, function(child) {
-          return persist(child)
-            .then(function(savedChild) {
-              children.push(savedChild);
-            });
-        }))
-        .then(function() {
-          return children;
+    function prepareRepeatedDocs(doc, repeated) {
+      if (repeated && repeated.child_data) {
+        return _.map(repeated.child_data, function(child) {
+          child.parent = doc;
+          return prepare(child);
         });
+      } else {
+        return [];
+      }
     }
 
-    function saveExtras(doc, original, extras) {
-      var children = [];
+    // Mutates the passed doc to attach prepared siblings, and returns all
+    // prepared siblings to be persisted
+    function prepareAndAttachSiblingDocs(doc, original, siblings) {
+      if (!doc._id) {
+        throw new Error('doc passed must already be prepared with an _id');
+      }
+
+      var preparedSiblings = [];
       var schema = $scope.unmodifiedSchema[doc.type];
 
-      // sequentially update all dirty db fields
-      var result = $q.resolve(doc);
-      _.each(schema.fields, function(conf, f) {
-        var customType = conf.type.match(/^(db|custom):(.*)$/);
+      _.each(schema.fields, function(fieldConf, fieldName) {
+        var customType = fieldConf.type.match(/^(db|custom):(.*)$/);
```

Comment:
So it's not actually a prefix, it's the entire matcher. The `(.*)` is used later to pull that value out.

I'm actually thinking this is easier to understand if you leave it there. Because it operates as both a matcher and an extractor of the second value, when you come across the code `preparedSibling.type = customType[2];` a little below, you can gather (if you know much about match) that the value is the bit after the `:`.